numba_cuda/numba/cuda/cudadrv/driver.py

```diff
@@ -2749,6 +2755,180 @@ def complete(self):
+
+        self.max_registers = max_registers
+        self.lineinfo = lineinfo
+        self.cc = cc
+        self.arch = arch
+        self.lto = lto
+        self.additional_flags = additional_flags
+
+        self._complete = False
+        self._object_codes = []
+        self.linker = None # need at least one program
+
+    @property
+    def info_log(self):
+        if not self.linker:
+            raise ValueError("Not Initialized")
+        return self.linker.get_info_log()
+
+    @property
+    def error_log(self):
+        if not self.linker:
+            raise ValueError("Not Initialized")
+        return self.linker.get_error_log()
+
+    def add_ptx(self, ptx, name='<cudapy-ptx>'):
+        obj = ObjectCode.from_ptx(ptx)
+        self._object_codes.append(obj)
+
+    def add_cu(self, cu, name='<cudapy-cu>'):
+        # TODO - vendor below logic somewhere common
+        cuda_include = [
+            get_cuda_paths()['include_dir'].info,
+        ]
+
+        cudadrv_path = os.path.dirname(os.path.abspath(__file__))
+        numba_cuda_path = os.path.dirname(cudadrv_path)
+
+        nrt_path = os.path.join(numba_cuda_path, "runtime")
+        include_paths = cuda_include + [numba_cuda_path, nrt_path]
+        prog = Program(
+            cu.decode('utf-8'),
+            'c++',
+            ProgramOptions(
+                arch=self.arch,
+                lineinfo=self.lineinfo,
+                max_register_count=self.max_registers,
+                relocatable_device_code=True,
+                include_path=include_paths,
+            )
+        )
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore",
+                message=(
+                    "The CUDA driver version is older"
+                )
+            )
+            obj = prog.compile('ptx')
+        self._object_codes.append(obj)
+        prog.close()
+
+    def add_cubin(self, cubin, name='<cudapy-cubin>'):
+        obj = ObjectCode.from_cubin(cubin)
+        self._object_codes.append(obj)
+
+    def add_ltoir(self, ltoir, name='<cudapy-ltoir>'):
+        obj = ObjectCode._init(ltoir, 'ltoir')
+        self._object_codes.append(obj)
+
+    def add_fatbin(self, fatbin, name='<cudapy-fatbin>'):
+        obj = ObjectCode._init(fatbin, 'fatbin')
+        self._object_codes.append(obj)
+
+    def add_ltoir(self, ltoir, name='<cudapy-ltoir>'):
+        obj = ObjectCode._init(ltoir, 'ltoir')
+        self._object_codes.append(obj)
+
+    def add_object(self, obj, name='<cudapy-object>'):
+        # TODO - hack
+        from cuda.bindings import nvjitlink
+        from cuda.core.experimental import _linker
+        _linker._nvjitlink_input_types['object'] = nvjitlink.InputType.OBJECT
+        ObjectCode._supported_code_type = (*ObjectCode._supported_code_type, 'object')
+        obj = ObjectCode._init(obj, 'object')
+        self._object_codes.append(obj)
+    
+
+    def add_library(self, lib, name='<cudapy-lib>'):
```

Comment:
There's a few cases in this class where I'm finding I have to do something like this. We discussed wanting to avoid having an `ObjectCode` constructor for every type of possible input, but it seems like numba requires the ability to assemble the eventual inputs to the link from a pretty broad variety of sources.

Reply:
It seems like we need an API to be able to construct `ObjectCode` instances from inputs of arbitrary kinds - archives, cubins, LTOIR, fatbins, etc.. It seems like the existing interface is aimed at providing something relatively safe, but doesn't allow for the lower-level / fine-grained control we need here to set up and complete a link.

Is that understanding correct?

Reply:
This is what I've arrived at doing, @leofang and I have had a few conversations about how this capability would fit into the broader goals of the `cuda-python` API, I'd be curious his thoughts.

Reply:
My question is if we really have to keep Numba's `Linker` class? Is it a public API that we must keep indefinitely?

We discussed in a past `cuda.core` design review meeting that Numba should just drop all of these linker classes once `cuda.core` is ready. In particular, we agreed that the `Linker` class can just take all inputs at once, instead of having all kinds of `add_<something>()` APIs to add things one by one. Do we really need to bend backwards here?

Reply:
xref: https://github.com/NVIDIA/cuda-python/issues/629